### PR TITLE
Add analyzer/codefixer to use AddAuthorizationBuilder

### LINKED ISSUE #45219: [Analyzer/Codefixer] Recommend using `AddAuthorizationBuilder` for configuring global policies
## Background and Motivation

In .NET 7, we introduced an `AddAuthorizationBuilder` extension method on `IServiceCollection` that would register authorization-related services and provide an `AuthorizationBuilder` for constructing policies. This is an abbreviated syntax that allows reduces nesting in the original pattern of calling `AddAuthorization` and providing a policy construct as a callback. 

### Analyzer Behavior and Message

When the user provides code where `AddAuthorizationBuilder` would provide a more abbreviated style, recommend a codefix with the following message:

> Use `AddAuthorizationBuilder` to register authorization services and construct policies.

### Category

- [ ] Design
- [ ] Documentation
- [ ] Globalization
- [ ] Interoperability
- [ ] Maintainability
- [ ] Naming
- [ ] Performance
- [ ] Reliability
- [ ] Security
- [ ] Style
- [X] Usage

### Severity Level

- [ ] Error
- [ ] Warning
- [X] Info
- [ ] Hidden

## Usage Scenarios
**Before**
```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddAuthorization(options =>
{
    options.AddPolicy("AtLeast21", policy =>
        policy.Requirements.Add(new MinimumAgeRequirement(21)));
});

var app = builder.Build();

app.UseAuthorization();

app.Run();
```
**After**
```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddAuthorizationBuilder()
  .AddPolicy("AtLeast21", policy =>
  {
        policy.Requirements.Add(new MinimumAgeRequirement(21)));
  });

var app = builder.Build();

app.UseAuthorization();

app.Run();
```

## Risks

Marking this as an info-only analyzer strikes a good balance between informing the user of this feature without being too presumptuous (via a warning). A refactoring would not have been good at educating the user about the functionality.

### PR DESCRIPTION
**Pull request by david-acker**
## Add `UseAuthorizationBuilderAnalyzer`, `AddAuthorizationBuilderFixer`

Adds an analyzer/codefixer to recommend that `AddAuthorization` be converted to `AddAuthorizationBuilder`.

### Description

The code fix converts any usage of the setters for the `DefaultPolicy`, `FaillbackPolicy`, and `InvokeHandlersAfterFailure` properties on `AuthorizationOptions` to an equivalent method call (`SetDefaultPolicy`, `SetFallbackPolicy`, and `SetInvokeHandlersAfterFailure`) on  `AuthorizationBuilder`.

No diagnostic is reported when the configure action passed to `AddAuthorization` uses the `GetPolicy(String)` method, the `DefaultPolicy` getter, the `FallbackPolicy` getter, or the `InvokeHandlersAfterFailure` getter on `AuthorizationOptions` as these do not exist on `AuthorizationBuilder` and thus cannot be converted.

No diagnostic is reported if the configure action passed to `AddAuthorization` contains operations unrelated to `AuthorizationOptions`, to avoid unintentionally deleting code when applying the code fix, as it would not be easy, or possible, to automatically map this to `AddAuthorizationBuilder`'s fluent API.

Fixes #45219

### CODE REVIEW

**Comment by mkArtakMSFT** [comment]
@halter73 can you please review this? Thanks!

File: src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddAuthorization(options =>
+{
+    _ = options.FallbackPolicy;
+    options.AddPolicy(""AtLeast21"", policy =>
```
Comment (halter73) [inline src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs:433]: I do like that this doesn't try to use the AuthorizationBuilder unless the configure AuthorizationOptions callback can be completely removed 👍

**Review by halter73 — COMMENTED** [review]
LGTM! Is there any particular reason this is still a draft? @captainsafia Can you take a look too?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        return SyntaxFactory.InvocationExpression(
+            SyntaxFactory.MemberAccessExpression(
+                SyntaxKind.SimpleMemberAccessExpression,
+                invocation.WithTrailingTrivia(
+                    SyntaxFactory.EndOfLine(Environment.NewLine),
+                    SyntaxFactory.Space,
+                    SyntaxFactory.Space,
+                    SyntaxFactory.Space,
+                    SyntaxFactory.Space),
```
Comment (halter73) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:123]: This part seems to be the sketchiest to me. I guess our tests expect `SyntaxFactory.EndOfLine("\r\n")`. Is this indentation relative to the previous line? Can we add tests for a more deeply nested call into `AddAuthorization(...)` like in `HostBuilder.ConfigureServices`?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                context.RegisterCodeFix(
+                    CodeAction.Create(title,
+                        cancellationToken => ReplaceWithAddAuthorizationBuilder(diagnostic, root, context.Document, invocation),
+                        equivalenceKey: title),
+                    diagnostic);
+            }
+        }
+    }
+
+    private static bool CanReplaceWithAddAuthorizationBuilder(Diagnostic diagnostic, SyntaxNode root, [NotNullWhen(true)] out InvocationExpressionSyntax? invocation)
```
[the remark this answers is not in the document]
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:53]: Offhand, I don't believe there should be any situations where a diagnostic would be emitted and this would return false. The analyzer should account for all of the scenarios where we wouldn't be able to apply the code fix and not emit a diagnostic in those situations. But I'll take a closer look to make sure the analyzer isn't missing anything.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 2 above)
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:123]: I agree. This part still felt pretty hacky to me, which is why I left the PR in draft. I haven't confirmed it yet, but I'm assuming this may also related to the test that's failing due to some line ending issues.

I added this in an attempt to improve the formatting of the code generated by the code fix so that it would match the example "after" code example in the issue, otherwise there wouldn't be any indentation relative to the previous line. 

Should we be concerned with the fixed code being "pretty" (within reason) or should we leave this up to the user to handle? If we do want to handle formatting as part of the code fix, I can go ahead and add tests for more deeply nested scenarios. Otherwise, I'll remove this and update the `fixedSource` values in the tests.

Fixed example: without the added indentation and newline
```csharp
builder.Services.AddAuthorizationBuilder().AddPolicy("AtLeast18", policy =>
        policy.Requirements.Add(new MinimumAgeRequirement(18))).AddPolicy("AtLeast21", policy =>
        policy.Requirements.Add(new MinimumAgeRequirement(21)));
```

File: src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs (hunk 1 above)
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs:433]: Thank you! I wanted to make the analyzer/code fix as "cautious" as possible, so it won't inadvertently break code that uses any `AuthorizationOptions`-specific APIs.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 2 above)
Comment (halter73) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:123]: I think we want it to look reasonably pretty if we're going to do this. The example without the added indentation and newline does look bad.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (halter73) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:53]: I'd just add an assert to document this if you're confident.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 2 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:123]: How helpful is the `SyntaxNode.NormalizeWhitespace` API here for formatting the `InvocationExpression`? I know it sometimes doesn't play too nicely with deeply-nested calls but wondering if it might help here.

Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:126]: Following up on this. Calling `NormalizeWhitespace` on the final `InvocationExpression` actually seems to format the block body examples OK, although the expression body examples ended up being pretty long since `NormalizeWhitespace` doesn't add new lines after the chained calls. So it looks like we'll need to add the `EndOfLineTrivia` ourselves if we want go this route.

Example: `MultipleAddPolicyCalls_UsingBlockBody_FixedWithAddAuthorizationBuilder`
```csharp
builder.Services.AddAuthorizationBuilder().AddPolicy("AtLeast18", policy =>
{
    policy.Requirements.Add(new MinimumAgeRequirement(18));
}).AddPolicy("AtLeast21", policy =>
{
    policy.Requirements.Add(new MinimumAgeRequirement(21));
});
```

Example: `MultipleAddPolicyCalls_UsingExpressionBody_FixedWithAddAuthorizationBuilder`
```csharp
builder.Services.AddAuthorizationBuilder().AddPolicy("AtLeast18", policy => policy.Requirements.Add(new MinimumAgeRequirement(18))).AddPolicy("AtLeast21", policy => policy.Requirements.Add(new MinimumAgeRequirement(21)));
```

Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs]: I also added two tests for the nested `AddAuthorization` scenarios and made some adjustments to have leading trivia is handled in `AddAuthorizationBuilderFixer` to ensure that indentation is correct.

**Comment by david-acker** [comment]
@halter73 @captainsafia How should we handle the line-ending issues with the Linux and MacOS test failures? Should we just add `OSSkipCondition` to these tests like we did for the line-ending issue with [Add analyzer and code fix to recommend against IHeaderDictionary.Add #44463](https://github.com/dotnet/aspnetcore/pull/44463)?

**Comment by halter73** [comment]
I didn't realize we're skipping tests due to line-ending issues. Ideally, we should split or normalize the line endings before asserting so developers can run the tests on any platform even if we don't expect platform-specific failures. We shouldn't assume everyone is developing on Windows.

We do this for our RequestDelegateGenerator tests.

https://github.com/dotnet/aspnetcore/blob/b4811e3612efae00ec6ef65b53163e47aa4e4f56/src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTestBase.cs#L319-L325

**Comment by david-acker** [comment]
How would you recommend hooking this into the current verification logic that's being used (from `Microsoft.CodeAnalysis.Testing`? At first glance, it looks like it may be possible to add in the line-ending normalization logic if we were to override `CodeFixTest.VerifyFixAsync`, although the source text verifications take place down in some private methods.

Also, as an aside, these tests did actually pass when I ran them locally on MacOS with a clean clone of this branch, oddly enough. I'm curious why this would be the case, since they seem to have failed consistently on CI.

**Comment by captainsafia** [comment]
@david-acker This issue with line-endings originally came up in https://github.com/dotnet/aspnetcore/pull/45642. At the time, we implemented the `ReplaceLineEndings` workaround in a single test in anticipation that we could move it lower down into the verifier code if it was encountered in more tests (like you have here). See [my comment here](https://github.com/dotnet/aspnetcore/pull/45642#issuecomment-1397491630) for where I think we should fix this.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        return false;
+    }
+
+    private static bool DoesOperationInvolveAuthorizationOptions(IOperation operation, AuthorizationOptionsTypes authorizationOptionsTypes)
+    {
+        if (operation is IExpressionStatementOperation { Operation: { } expressionStatementOperation })
+        {
+            if (expressionStatementOperation is ISimpleAssignmentOperation { Target: IPropertyReferenceOperation { Property.ContainingType: { } propertyReferenceContainingType } }
```
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs:92]: Do we need add tests that what these code segments are checking (assuming I understood the code correctly) like:

```csharp
builder.Services.AddAuthorization(options => ConfigureOptions(options));
```

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+        var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+        if (semanticModel == null)
+        {
+            return;
+        }
+
+        foreach (var diagnostic in context.Diagnostics)
+        {
+            if (CanReplaceWithAddAuthorizationBuilder(diagnostic, root, out var invocation))
```
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:43]: Yay! Thanks for doing the feasibility check before registering the codefix. This was an issue in the first few codefixers that we did...

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            {
+                const string title = "Use 'AddAuthorizationBuilder'";
+                context.RegisterCodeFix(
+                    CodeAction.Create(title,
+                        cancellationToken => ReplaceWithAddAuthorizationBuilder(diagnostic, root, context.Document, invocation),
+                        equivalenceKey: title),
```
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:49]:
```suggestion
                        equivalenceKey: EquivalencyKey.Create(DiagnosticDescriptors.UseAddAuthorizationBuilder.Id)),
```

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        InvocationExpressionSyntax invocation,
+        string invokedMemberName,
+        ArgumentListSyntax argumentList)
+    {
+        var invocationLeadingTrivia = invocation.GetLeadingTrivia()
+            .Where(trivia => !trivia.IsKind(SyntaxKind.EndOfLineTrivia));
+        var newInvocationTrivia = new SyntaxTriviaList(
+            SyntaxFactory.EndOfLine(
+                Environment.NewLine),
+                SyntaxFactory.Space,
```
Comment (captainsafia) [inline src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:124]: QQ: is this the only way to create tabs with the SyntaxFactory?

**Review by captainsafia — COMMENTED** [review]
Looking good! Thanks for the `ReplaceLineEndings` fixup. Left some comments and questions inline.

File: src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+            .WithMessage(Resources.Analyzer_UseAddAuthorizationBuilder_Message);
+
+        var source = @"
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+
+var builder = WebApplication.CreateBuilder(args);
+
+{|#0:builder.Services.AddAuthorization(options =>
```
[the remark this answers is not in the document]
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs]: Ah interesting point, I hadn't considered how named parameters could impact this originally. Updated!

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        return invocation.Parent is IExpressionStatementOperation;
+    }
+
+    private static void AddDiagnosticInformation(OperationAnalysisContext context, Location location)
+    {
+        context.ReportDiagnostic(Diagnostic.Create(
+            DiagnosticDescriptors.UseAddAuthorizationBuilder,
+            location));
+    }
+    private sealed class AuthorizationOptionsTypes
```
[the remark this answers is not in the document]
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs]: Addressed!

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 6 above)
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:49]: Hmm, I'm not seeing anything for this `EquivalencyKey` class. Where would I be able to find this?

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 7 above)
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs:124]: I've updated this to use `SyntaxFactory.Tab`, which seems like a better fit.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 6 above)
Comment (captainsafia) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs]: Total brain fart on my part. I think we can just do `equivalenceKey : DiagnosticDescriptors.UseAddAuthorizationBuilder.Id`.

I notice that there are some places in our code where we use the title as the equivalence key and others where we use the diagnostic ID, so it appears we're inconsistent in our approach here.

Based on the docs [here](https://github.com/dotnet/roslyn/blob/main/docs/analyzers/FixAllProvider.md#selecting-an-equivalence-key-for-code-actions), using the title is recommended for most cases unless you have a fixer that targets different values.

TL;DR: Disregard my feedback here!

File: src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+        services.AddAuthorizationBuilder()
+            .AddPolicy(""AtLeast21"", policy =>
+                policy.Requirements.Add(new MinimumAgeRequirement(21)));
+    });
+";
+
+        await VerifyCodeFix(source, new[] { diagnostic }, fixedSource);
+    }
+
+    private static async Task VerifyCodeFix(string source, DiagnosticResult[] diagnostics, string fixedSource)
```
[the remark this answers is not in the document]
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/test/Authorization/AddAuthorizationBuilderTests.cs:728]: Added `AddAuthorization_CallAssignedToVariable_NoDiagnostic`.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs (hunk 4 above)
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs:92]: I've added some new tests and updated a few of the existing ones to ensure that all of the branches in these private methods are covered by tests. I also removed the `MethodReferenceOperation`-related branch as this was unreachable.

File: src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs (hunk 6 above)
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/CodeFixes/Authorization/AddAuthorizationBuilderFixer.cs]: Good to know! I've updated this to use the DiagnosticDescriptor Id as the equivalenceKey.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                AddDiagnosticInformation(context, invocation.Syntax.GetLocation());
+            }
+
+        }, OperationKind.Invocation);
+    }
+
+    private static bool IsCompatibleWithAuthorizationBuilder(IInvocationOperation invocation, AuthorizationOptionsTypes authorizationOptionsTypes)
+    {
+        if (invocation is { Arguments: { Length: 2 } invocationArguments }
```
[the remark this answers is not in the document]
Comment (david-acker) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Authorization/AddAuthorizationBuilderAnalyzer.cs]: I've refactored each of these pattern checks into a separate private method to make it clearer what each one does. Right now I've replaced each check with it's respective method to keep things as "flat" as possible, although this does require these methods to effectively be chained with their out parameters. 

Let me know if there are any additional improvements or changes you would suggest here!

**Review by captainsafia — APPROVED** [review]
LGTM! :shipit:
